xdsl/dialects/gpu.py

```diff
@@ -109,6 +109,55 @@ def print_parameters(self, printer: Printer) -> None:
 DimensionAttr = _GPUAttr[_DimensionAttr]
 AllReduceOperationAttr = _GPUAttr[_AllReduceOperationAttr]
 
+_Element = TypeVar("_Element", bound=Attribute, covariant=True)
+
+
+@irdl_op_definition
+class AllocOp(Operation):
+    name = "gpu.alloc"
+    hostShared: OptOpAttr[UnitAttr]
+    asyncDependencies: Annotated[VarOperand, AsyncTokenType]
+    dynamicSizes: Annotated[VarOperand, IndexType]
+    symbolOperands: Annotated[VarOperand, IndexType]
+
+    irdl_options = [AttrSizedOperandSegments()]
+
+    memref: Annotated[OpResult, memref.MemRefType[Attribute]]
+    asyncToken: Annotated[OptOpResult, AsyncTokenType]
+
+    def verify_(self) -> None:
+        ndyn = len(self.dynamicSizes)
+        assert isinstance(self.memref.typ, memref.MemRefType)
+        typ: memref.MemRefType[Attribute] = self.memref.typ
+        ndyn_typ = len([i for i in typ.shape.data if i.value.data == -1])
+        if ndyn != ndyn_typ:
+            raise VerifyException(
+                f"Expected {ndyn_typ} dynamic sizes, got {ndyn}. All "
+                "dynamic sizes need to be set in the alloc operation.")
+
+    @staticmethod
+    def get(return_type: memref.MemRefType[_Element],
+            dynamic_sizes: Sequence[SSAValue | Operation]
+            | None = None,
+            host_shared: bool = False,
+            async_dependencies: Sequence[SSAValue | Operation] | None = None,
+            is_async: bool = False) -> AllocOp:
+        token_return = [AsyncTokenType()] if is_async else []
+        dynamic_sizes_vals: list[SSAValue] = [] if dynamic_sizes is None else [
```

Comment:
Ooh, me too actually! Forgot about the simple "if <object>", so I tried to avoid the `[...] if dynamic_sizes is not None else []` which just hurst my brain in layered negations. Thanks for the reminder!